### PR TITLE
feat(nodebuilder/p2p): Implement `Module`

### PR DESCRIPTION
**Pull request by renaynay**
Implements the P2P `Module` and provides it on our RPC.

Fixes one TODO in https://github.com/celestiaorg/celestia-node/issues/977

### CODE REVIEW

File: nodebuilder/p2p/p2p.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	BandwidthStats       func() metrics.Stats
+	BandwidthForPeer     func(id peer.ID) metrics.Stats
+	BandwidthForProtocol func(proto protocol.ID) metrics.Stats
+	PubSubPeers          func(topic string) []peer.ID
+}
+
+type Module interface {
+	// Info returns basic information about the node's p2p host/operations.
+	Info() Info
+	// Peers returns all peer IDs used across all inner stores. // TODO should we
```
[the remark this answers is not in the document]
Comment (renaynay) [reply nodebuilder/p2p/p2p.go]: done

File: nodebuilder/tests/rpc_test.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package tests
+
+/*
+func TestRPC(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (renaynay) [reply nodebuilder/tests/rpc_test.go:4]: deleted as we'll implement rpc tests via integration later once we figure out the best structure for them.

File: nodebuilder/p2p/p2p.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+// Info contains basic information about the node's p2p host/operations.
+type Info struct {
+	// ID is the node's peer ID
+	ID peer.ID `json:"id"`
+	// Addrs is the node's
+	Addrs []ma.Multiaddr `json:"addrs"`
+}
+
+func (m *manager) Info() Info {
+	m.host.Network().
```
Comment (vgonkivs) [inline nodebuilder/p2p/p2p.go:136]: ?

**Comment by renaynay** [comment]
Use assert

File: nodebuilder/p2p/p2p.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	}
+}
+
+// Info contains basic information about the node's p2p host/operations.
+type Info struct {
+	// ID is the node's peer ID
+	ID peer.ID `json:"id"`
+	// Addrs is the node's
+	Addrs []ma.Multiaddr `json:"addrs"`
+}
```
Comment (renaynay) [inline nodebuilder/p2p/p2p.go:133]: Not really sure what else is important to return @Wondertan

Comment (distractedm1nd) [reply nodebuilder/p2p/p2p.go:133]: we can add more if we think of any. would merge with this as is

**Review by distractedm1nd — APPROVED** [review]

File: nodebuilder/p2p/p2p.go (hunk 4 above)
Comment (renaynay) [reply nodebuilder/p2p/p2p.go:134]: agreed

File: nodebuilder/p2p/p2p_test.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/p2p/host/autonat"
+	mocknet "github.com/libp2p/go-libp2p/p2p/net/mock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestP2PModule_methodsOnHost tests P2P Module methods on
+// the instance of Host.
+func TestP2PModule_methodsOnHost(t *testing.T) {
```
Comment (distractedm1nd) [inline nodebuilder/p2p/p2p_test.go:24]: [non-blocking] optional nit: i dont like the naming scheme of _methodsOnHost. I think TestP2PModule_Host would be better

File: nodebuilder/rpc/rpc.go [hunk 6]
```diff
@@ -16,13 +17,15 @@ func RegisterEndpoints(
...
 	daser das.Module,
+	p2p p2p.Module,
 	serv *rpc.Server,
 ) {
 	serv.RegisterService("state", state)
 	serv.RegisterService("share", share)
-	serv.RegisterService("fraud", fraud)
 	serv.RegisterService("header", header)
 	serv.RegisterService("das", daser)
+	serv.RegisterService("fraud", fraud)
```
Comment (distractedm1nd) [inline nodebuilder/rpc/rpc.go:27]: why did you move fraud

**Review by distractedm1nd — APPROVED** [review]

File: nodebuilder/rpc/rpc.go (hunk 6 above)
Comment (renaynay) [reply nodebuilder/rpc/rpc.go:27]: relic

**Review by distractedm1nd — APPROVED** [review]

**Review by Wondertan — CHANGES_REQUESTED** [review]
Requesting some changes

File: nodebuilder/p2p/p2p_test.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+	require.NoError(t, err)
+
+	mgr := newManager(host, nil, nat, nil, nil, nil)
+
+	assert.Equal(t, nat.Status(), mgr.NATStatus())
+}
+
+// TestP2PModule_Bandwidth tests P2P Module methods on
+// the Host's bandwidth reporter.
+func TestP2PModule_Bandwidth(t *testing.T) {
```
Comment (renaynay) [inline nodebuilder/p2p/p2p_test.go:85]: this test will likely be flakey

File: nodebuilder/p2p/resource.go [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package p2p
+
+import (
+	"github.com/libp2p/go-libp2p-core/network"
+	rcmgr "github.com/libp2p/go-libp2p-resource-manager"
+)
+
+func ResourceManager() (network.ResourceManager, error) {
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply nodebuilder/p2p/resource.go:8]: @renaynay

File: nodebuilder/p2p/p2p_test.go (hunk 7 above)
Comment (Wondertan) [reply nodebuilder/p2p/p2p_test.go:83]: If so, we can increase the sleep below

File: nodebuilder/p2p/p2p_test.go [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+// the node's instance of AutoNAT.
+func TestP2PModule_Autonat(t *testing.T) {
+	host, err := libp2p.New(libp2p.EnableNATService())
+	require.NoError(t, err)
+
+	mgr := newManager(host, nil, nil, nil, nil)
+
+	status, err := mgr.NATStatus()
+	assert.NoError(t, err)
+	// TODO ??????
```
Comment (Wondertan) [inline nodebuilder/p2p/p2p_test.go:77]: ?

File: nodebuilder/p2p/p2p.go [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+	ResourceState() (rcmgr.ResourceManagerStat, error)
+
+	// PubSubPeers returns the peer IDs of the peers joined on
+	// the given topic.
+	PubSubPeers(topic string) []peer.ID
+}
+
+// manager contains all components necessary to access information and
+// perform actions related to the node's p2p Host / operations.
+type manager struct {
```
[the remark this answers is not in the document]
Comment (renaynay) [reply nodebuilder/p2p/p2p.go:105]: fine too, originally had a todo here that considered changing name.

File: nodebuilder/p2p/p2p.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	rcmgr "github.com/libp2p/go-libp2p-resource-manager"
+	basichost "github.com/libp2p/go-libp2p/p2p/host/basic"
+	"github.com/libp2p/go-libp2p/p2p/net/conngater"
+)
+
+// API is a wrapper around Module for the RPC.
+// TODO(@distractedm1nd): These structs need to be autogenerated.
+//
+//nolint:dupl
```
Comment (distractedm1nd) [inline nodebuilder/p2p/p2p.go:22]: did you ever figure out why this is `dupl` and the others aren't?

File: nodebuilder/p2p/p2p.go [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+	}
+}
+
+// Info contains basic information about the node's p2p host/operations.
+type Info struct {
+	// ID is the node's peer ID
+	ID peer.ID `json:"id"`
+	// AddrInfo is the node's
+	AddrInfo peer.AddrInfo `json:"addr_info"`
+}
```
Comment (Wondertan) [inline nodebuilder/p2p/p2p.go:135]: AddrInfo already contains the peer.ID, so we simply return the peer.AddrInfo instead of our custom struct

**Review by Wondertan — CHANGES_REQUESTED** [review]
Almost there

File: nodebuilder/p2p/p2p.go (hunk 12 above)
Comment (Wondertan) [reply nodebuilder/p2p/p2p.go:135]: Or we can keep a struct, but remove the peer.ID, so if we ever want to add a new field, it won't be breaking

File: nodebuilder/p2p/p2p.go [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+	ListBlockedPeers     func() []peer.ID
+	Protect              func(id peer.ID, tag string)
+	Unprotect            func(id peer.ID, tag string) bool
+	IsProtected          func(id peer.ID, tag string) bool
+	BandwidthStats       func() metrics.Stats
+	BandwidthForPeer     func(id peer.ID) metrics.Stats
+	BandwidthForProtocol func(proto protocol.ID) metrics.Stats
+	ResourceState        func() (rcmgr.ResourceManagerStat, error)
+	PubSubPeers          func(topic string) []peer.ID
+}
```
Comment (Wondertan) [inline nodebuilder/p2p/p2p.go:42]: Readability nit: let's keep this auxiliary API struct somewhere below of the file, while Module the interface at the top

Comment (renaynay) [reply nodebuilder/p2p/p2p.go:42]: kk

**Review by Wondertan — APPROVED** [review]
Blessed

**Review by distractedm1nd — APPROVED** [review]
blessed + based

**Review by liamsi — APPROVED** [review]
ADR changes are tiny and make sense
